Sources/AsyncHTTPClient/ConnectionPool/HTTPConnectionPool+Factory.swift

```diff
@@ -0,0 +1,314 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2021 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Logging
+import NIO
+import NIOHTTP1
+import NIOSSL
+import NIOTLS
+#if canImport(Network)
+    import NIOTransportServices
+#endif
+
+extension HTTPConnectionPool {
+    enum NegotiatedProtocol {
+        case http1_1(Channel)
+        case http2_0(Channel)
+    }
+
+    final class ConnectionFactory {
```

Comment:
Does this need to be a `class`?

Reply:
Same question

Reply:
In general I'm inclined to make things values when they can be.

Reply:
We can save a Heap allocation by making it a struct. It probably doesn't matter, but I'd say it makes sense to use a `struct` in this case.

Reply:
We capture this factory in _a lot_ of EventLoopFuture `flatMap`s. Yes we save one heap allocation per connection pool, but we will start to copy around four config properties for every closure. Of course we could encapsulate the config in a storage class to offset the copying... But is this really worth here?

Reply:
I think this represents a premature optimisation. So far I've only seen a few flatMaps where this is the _only_ object captured (i.e. where we'd remove an allocation by having this be a `class`). We can reduce the copying if needed by extracting to temporaries and only closing over the specific thing we need.

In general I care much more about getting semantics right than about initial performance numbers. As you say, we can always refactor to a class-backed struct if we want the performance profile of a `class`. The question is, conceptually, _should_ this object be a value? In general I'm inclined to make objects that should be values into values.

Reply:
I agree. Thanks for the explanation. Fixed.